python/src/trezorlib/ethereum.py

```diff
@@ -17,11 +17,157 @@
 from . import exceptions, messages
 from .tools import expect, normalize_nfc, session
 
+from eth_abi.packed import encode_single_packed
+import json
+import re
 
 def int_to_big_endian(value):
     return value.to_bytes((value.bit_length() + 7) // 8, "big")
 
 
+type_name_re = re.compile("^\\w*")
+
+
+def find_typed_dependencies(primary_type: str, types: dict, results: list = None):
+    """
+    Finds all types within a type definition object
+
+    primary_type - Root type
+    types - Type definitions
+    results - Current set of accumulated types
+    """
+    if results is None:
+        results = []
+
+    m = type_name_re.match(primary_type)
+    if m:
+        primary_type = m.string[m.start():m.end()]
+    else:
+        raise ValueError("cannot parse primary type: %s" % primary_type)
+
+    if (primary_type in results) or (types.get(primary_type) is None):
+        return results
+
+    results = results + [primary_type]
+    for field in types[primary_type]:
+        deps = find_typed_dependencies(field["type"], types, results)
+        for dep in deps:
+            if not dep in results:
+                results = results + [dep]
+
+    return results
+
+def encode_type(primary_type: str, types: dict):
+    """
+    Encodes the type of an object by encoding a comma delimited list of its members
+
+    primary_type - Root type to encode
+    types - Type definitions
+    """
+    result = ""
+    result_indexed = {}
+
+    deps = find_typed_dependencies(primary_type, types)
+    deps = list(filter(lambda dep: dep != primary_type, deps))
+    deps = [primary_type] + sorted(deps)
+
+    for type_name in deps:
+        children = types.get(type_name)
+        if children is None:
+            raise ValueError("no type definition specified: %s" % type_name)
+        fields = ",".join(map(lambda field: "%s %s" % (field["type"], field["name"]), children))
+        result_indexed[type_name] = [field for (_, field) in enumerate(children)]
+        result += "%s(%s)" % (type_name, fields)
+
+    return result, result_indexed
+
+allowed_typed_data_properties = ["types", "primaryType", "domain", "message"]
+
+def sanitize_typed_data(data: dict):
+    """
+    Removes properties from a message object that are not defined per EIP-712
+
+    data - typed message object
+    """
+    sanitized_data = {}
+    for key in allowed_typed_data_properties:
+        val = data.get(key)
+        if val is None:
+            continue
+        sanitized_data[key] = val
+
+    if "types" in sanitized_data:
+        sanitized_data["types"] = { "EIP712Domain": [], **sanitized_data["types"] }
```

Comment:
I don't get the problem here. This line was a carbon copy of this one:
https://github.com/MetaMask/eth-sig-util/blob/41a62c8255ce061130f85876be178e65e6ae99a2/src/index.ts#L286

I tried to stay close to that implementation. Isn't it correct?

Reply:
I'm not familiar with TypeScript, the translation is probably alright, just very unusual to write it like this in python.

Reply:
that might be idiomatic typescript (presumably? i'm not fluent) but it's decidedly not idiomatic Python. let's go with @mmilata's suggestion instead

Reply:
It seems this line has a meaning in the way it is written - we want to make sure the `EIP712Domain` key is always there and defaulting to an empty list. When it is already present in `sanitized_data["types"]`, it will then overwrite this empty list with current value (therefore the dict unpacking with `**`).

What is the best way of handling this situation? We could do an explicit check for this:
```python
if "EIP712Domain" not in sanitized_data["types"]:
    sanitized_data["types"]["EIP712Domain"] = []
```

When using only `sanitized_data["types"]["EIP712Domain"] = []`, we received `FirmwareError (99)` - most probably is complaining about the `EIP712Domain` being empty (it probably shouldn't though)

Reply:
Changed in a05275a